---
title: "BLAST Analysis"
output: html_notebook
---

### BLASTN
Align Dmel cDNA to species genome
```{r}
library(ggplot2)
library(dplyr)
library(UpSetR)
library(biomaRt)

url.prefix <- 'https://raw.githubusercontent.com/manzt/gi_a2/master/blastn-results/'
species <- c('dgri', 'dmoj', 'dvir')

cols <- c('qseqid', 'chrom', 'pident', 'length', 'mismatch', 'gapopen', 
             'qstart', 'qend', 'start', 'end', 'evalue', 'bitscore')

# aggregate blast results in dataframe
raw_data <- list()
for (s in species) {
  url <- paste0(url.prefix, s, '_blastn.txt')
  print(paste('Fetching', s, 'data from:', url))
  raw_data[[s]] <- read.csv(url, col.names=cols) %>% mutate(species=s)
}
blastn <- bind_rows(raw_data)
blastn
```
```{r}
# total blast alignments for each species
blastn %>% 
  group_by(species) %>% 
  tally()
```
```{r}
# define e-value threshold and filter data
thresh <- 150
filtered <- blastn %>% filter(bitscore > thresh)

# total blast alignments for each species after thresh filter
filtered %>% 
  group_by(species) %>% 
  tally()
```

```{r}
# count transcripts found from Dmel 
found_genes <- filtered %>%
  group_by(species, qseqid) %>%
  summarise()

# number of Dmel genes in each
found_genes %>% tally()
```
```{r}
# plot Intersection
cdna.by_species <- lapply(species, function(s) found_genes$qseqid[found_genes$species == s])
names(cdna.by_species) <- species
upset(fromList(cdna.by_species), order.by='freq')
```

```{r}
# Helper function - returns subsets used for upset plot
# get overlapping subsets from upsetR plots
# https://github.com/hms-dbmi/UpSetR/issues/85#issuecomment-415480954
overlapGroups <- function (listInput, sort = TRUE) {
  listInputmat    <- fromList(listInput) == 1
  listInputunique <- unique(listInputmat)
  grouplist <- list()
  # going through all unique combinations and collect elements for each in a list
  for (i in 1:nrow(listInputunique)) {
    currentRow <- listInputunique[i,]
    myelements <- which(apply(listInputmat,1,function(x) all(x == currentRow)))
    attr(myelements, "groups") <- currentRow
    grouplist[[paste(colnames(listInputunique)[currentRow], collapse = ":")]] <- myelements
    myelements
  }
  if (sort) {
    grouplist <- grouplist[order(sapply(grouplist, function(x) length(x)), decreasing = TRUE)]
  }
  attr(grouplist, "elements") <- unique(unlist(listInput))
  return(grouplist)
  # save element list to facilitate access using an index in case rownames are not named
}
```

```{r} 
# create table of molecular function GO-ids 
gene_ontology <- tibble(
  go_id=c('GO:0005215', 'GO:0045182', 'GO:0003824', 'GO:0016247', 'GO:0004872', 
    'GO:0004871', 'GO:0016209', 'GO:0005198', 'GO:0005488'),
  description=c('transporter activity', 'translation regulator activity', 
    'catalytic activity', 'channel regulator activity', 
    'receptor activity', 'signal transducer activity', 
    'antioxidant activity', 'structural molecule activity', 'binding')
)
gene_ontology
```
```{r} 
bio_process <- tibble(
  go_id=c('GO:0071840', 'GO:0009987', 'GO:0051179', 'GO:0000003',
    'GO:0065007', 'GO:0050896', 'GO:0032502', 'GO:0048511', 
    'GO:0032501', 'GO:0008152', 'GO:0040007'),
  description=c('cellular component organization or biogenesis',
    'cellular process', 'localization', 'reproduction', 
    'biological regulation', 'response to stimulus',
    'developmental process', 'rhythmic process', 
    'multicellular organismal process', 
    'metabolic process', 'growth')
)
bio_process
```

```{r }
# initialize biomaRt to query go_ids
ensembl <- useMart(biomart="ENSEMBL_MART_ENSEMBL", 
                   dataset="dmelanogaster_gene_ensembl", 
                   host="www.ensembl.org")
```

```{r }
# Query ensembl for go_ids and refine to molecular function ids for each set
sets <- overlapGroups(cdna.by_species)
data_list <- list()
for (i in 1:length(sets)) {
  transcript_ids <- attr(sets, 'elements')[sets[[i]]]
  data_list[[i]] <- getBM(attributes = c('ensembl_gene_id', 'go_id', 'namespace_1003'),
                     filters='ensembl_transcript_id',
                     values=transcript_ids, mart=ensembl) %>% 
    filter(go_id %in% bio_process$go_id) %>% 
    mutate(set=names(sets[i]))
}
set_ids <-  bind_rows(data_list) %>% left_join(bio_process, by='go_id') 
set_ids
```

```{r}
ggplot(set_ids, aes(x=set, fill=description)) + 
  geom_bar(width = 0.5, color='black') + 
  scale_x_discrete(limits=names(sets)) + 
  theme_classic() + 
  theme(legend.position="bottom") + 
  labs(x='', y='', fill='')
```

```{r} 
set_ids %>% group_by(set) %>% count()
```

### TBLASTN
Align Dmel pepdies to species genome
```{r}
paths <- c('/mhome/maths/q/tjm74/gi/assignment_2/dgri_tblastn.txt',
           '/mhome/maths/t/ldk26/private/GenomeInformatics/D.mojavensis/result_tblastn2.out',
           '/mhome/maths/s/as2841/GIA2/dvri_tblastn.txt')

species <- c('dgri', 'dmoj', 'dvir')

dgri.cols <- c('qseqid', 'sseqid', 'pident', 'length', 'mismatch', 'gapopen', 
          'qstart', 'qend', 'sstart', 'send', 'evalue', 'bitscore')

dmoj.cols <- c('qseqid', 'sseqid', 'pident', 'length', 'qstart', 
               'sstart', 'send', 'evalue', 'bitscore')

dvir.cols <- c('qseqid', 'sseqid', 'pident', 'length', 'mismatch', 'gapopen', 
              'qstart', 'qend', 'sstart', 'send', 'evalue', 'bitscore')

cols <- list(dgri.cols, dmoj.cols, dvir.cols)

raw_data <- list()
for (i in 1:3) {
  print(paths[i])
  raw_data[[species[i]]] <- read.csv(paths[i], col.names=cols[[i]]) %>% mutate(species=species[i])
  head(raw_data[[species[i]]])
}
tblastn <- bind_rows(raw_data)
tblastn
```

```{r }
# total blast alignments for each species
tblastn %>% 
  group_by(species) %>% 
  tally()
```

```{r }
thresh.t <- 300
tblast.filtered <- tblastn %>%
  filter(bitscore > thresh.t)

tblast.filtered %>% group_by(species) %>% count()
```

```{r}
# count proteins found from Dmel 
found_peps <- tblast.filtered %>%
  group_by(species, qseqid) %>%
  summarise()

# number of Dmel genes in each
found_peps %>% tally()
```

```{r}
# plot Intersection
pep.by_species <- lapply(species, function(s) found_peps$qseqid[found_peps$species == s])
names(pep.by_species) <- species
pdf('upset-tblastn.pdf', width=7, height=5)
upset(fromList(pep.by_species), order.by='freq')
dev.off()
upset(fromList(pep.by_species), order.by='freq')

```


```{r }
# Query ensembl for go_ids and refine to molecular function ids for each set
pep.sets <- overlapGroups(pep.by_species)
data_list <- list()
for (i in 1:length(pep.sets)) {
  gene_ids <- attr(pep.sets, 'elements')[pep.sets[[i]]]
  data_list[[i]] <- getBM(attributes = c('ensembl_gene_id', 'go_id', 'namespace_1003'),
                     filters='flybase_translation_id',
                     values=gene_ids, mart=ensembl) %>% 
    filter(go_id %in% bio_process$go_id) %>% 
    mutate(set=names(pep.sets[i]))
}
pep.set_ids <-  bind_rows(data_list) %>% left_join(bio_process, by='go_id') 
pep.set_ids
```

```{r}
ggplot(pep.set_ids, aes(x=set, fill=description)) + 
  geom_bar(width = 0.5, color='black') + 
  scale_x_discrete(limits=names(sets)) + 
  theme_classic() + 
  theme(legend.position="bottom") + 
  labs(x='', y='', fill='')
```

### Compare to Model
```{r}

```



```{r sushi attempt}
# library(Sushi)
# library(valr)

chrom = 'scaffold_15110'
chromstart = 0
chromend = 24565398 - 20000000
blastn %>%
  filter(chrom == chrom) %>%
  filter(start < chromend) %>%
  ggplot(aes(x=start, width=length, y=species)) + geom_tile()


# merge together nearby alignments
allowed_spacing <- 1.5e6
bed_cluster(blastn %>% group_by(qseqid, chrom), max_dist=allowed_spacing) %>% 
  filter(qseqid == 'FBtr0334137' & chrom == 'scaffold_15126') %>%
  View()

x <- bed_merge(blastn %>% filter(evalue < 1E-50) %>% group_by(qseqid, chrom, species), max_dist=allowed_spacing)
x
```
